### DISCUSSION TITLE
Link failed with latest GitHub source of MAME

### DISCUSSION THREAD

**Discussion by wyatt-wong** [mamedev/discussions#140] [General]
I downloaded the latest GitHub source of MAME using the following command in macOS Sonoma 14.6.1

`https://github.com/mamedev/mame`

However, whenever I build the source, I received the following error:

`wyattwong@MacBookPro mame % make
Clang 15.0.0 detected
make -R --no-print-directory -C build/projects/sdl/mame/gmake-osx-clang config=release64 precompile
Compiling src/mame/igs/igs_fear.cpp...
Compiling src/mame/igs/igs_m027.cpp...
Compiling src/mame/igs/igs_m027xa.cpp...
Archiving libigs.a...
Compiling src/devices/cpu/xa/xa.cpp...
Archiving liboptional.a...
Compiling generated/version.cpp...
Emitting mame_mame-Info.plist...
Linking mame...
Undefined symbols for architecture x86_64:
  "_CONCEPT_KEYBOARD", referenced from:
      concept_state::corvus_concept(machine_config&) in libconcept.a[2](concept.o)
  "_IGS027A", referenced from:
      (anonymous namespace)::igs_fear_state::igs_fear(machine_config&) in libigs.a[14](igs_fear.o)
      (anonymous namespace)::igs_m027_state::m027(machine_config&) in libigs.a[15](igs_m027.o)
      (anonymous namespace)::igs_m027xa_state::igs_mahjong_xa(machine_config&) in libigs.a[17](igs_m027xa.o)
  "_IGS_XA_ICS_SOUND", referenced from:
      (anonymous namespace)::igs_fear_state::igs_fear(machine_config&) in libigs.a[14](igs_fear.o)
  "_IGS_XA_SUBCPU", referenced from:
      (anonymous namespace)::igs_m027xa_state::igs_mahjong_xa(machine_config&) in libigs.a[17](igs_m027xa.o)
  "igs027a_cpu_device::trigger_irq(unsigned int)", referenced from:
      (anonymous namespace)::igs_fear_state::sound_irq(int) in libigs.a[14](igs_fear.o)
      (anonymous namespace)::igs_m027_state::interrupt(timer_device&, int) in libigs.a[15](igs_m027.o)
      (anonymous namespace)::igs_m027xa_state::xa_irq(int) in libigs.a[17](igs_m027xa.o)
      (anonymous namespace)::igs_m027xa_state::interrupt(timer_device&, int) in libigs.a[17](igs_m027xa.o)
  "igs027a_cpu_device::igs027a_cpu_device(machine_config const&, char const*, device_t*, unsigned int)", referenced from:
      auto machine_config::device_add<emu::detail::device_type_impl<igs027a_cpu_device> const&, int>(char const*, emu::detail::device_type_impl<igs027a_cpu_device> const&, int&&) in libigs.a[14](igs_fear.o)
  "igs_xa_mcu_device_base::cmd_w(unsigned short)", referenced from:
      (anonymous namespace)::igs_fear_state::main_map(address_map&) in libigs.a[14](igs_fear.o)
      (anonymous namespace)::igs_m027xa_state::main_map(address_map&) in libigs.a[17](igs_m027xa.o)
  "concept_keyboard_device::write_rxd(int)", referenced from:
      concept_state::corvus_concept(machine_config&) in libconcept.a[2](concept.o)
  "concept_keyboard_device::concept_keyboard_device(machine_config const&, char const*, device_t*, unsigned int)", referenced from:
      auto machine_config::device_add<emu::detail::device_type_impl<concept_keyboard_device> const&>(char const*, emu::detail::device_type_impl<concept_keyboard_device> const&) in libconcept.a[2](concept.o)
  "igs_xa_mcu_subcpu_device::irqack_w(unsigned short)", referenced from:
      (anonymous namespace)::igs_m027xa_state::main_map(address_map&) in libigs.a[17](igs_m027xa.o)
  "igs_xa_mcu_subcpu_device::igs_xa_mcu_subcpu_device(machine_config const&, char const*, device_t*, unsigned int)", referenced from:
      auto machine_config::device_add<emu::detail::device_type_impl<igs_xa_mcu_subcpu_device> const&, int>(char const*, emu::detail::device_type_impl<igs_xa_mcu_subcpu_device> const&, int&&) in libigs.a[17](igs_m027xa.o)
  "igs_xa_mcu_ics_sound_device::igs_xa_mcu_ics_sound_device(machine_config const&, char const*, device_t*, unsigned int)", referenced from:
      auto machine_config::device_add<emu::detail::device_type_impl<igs_xa_mcu_ics_sound_device> const&, int>(char const*, emu::detail::device_type_impl<igs_xa_mcu_ics_sound_device> const&, int&&) in libigs.a[14](igs_fear.o)
  "typeinfo for igs_xa_mcu_device_base", referenced from:
      util::detail::delegate_generic_class* util::detail::delegate_late_bind_helper<delegate_late_bind>::late_bind_helper<igs_xa_mcu_device_base>(delegate_late_bind&) in libigs.a[14](igs_fear.o)
      util::detail::delegate_generic_class* util::detail::delegate_late_bind_helper<delegate_late_bind>::late_bind_helper<igs_xa_mcu_device_base>(delegate_late_bind&) in libigs.a[14](igs_fear.o)
  "typeinfo for concept_keyboard_device", referenced from:
      util::detail::delegate_generic_class* util::detail::delegate_late_bind_helper<delegate_late_bind>::late_bind_helper<concept_keyboard_device>(delegate_late_bind&) in libconcept.a[2](concept.o)
      util::detail::delegate_generic_class* util::detail::delegate_late_bind_helper<delegate_late_bind>::late_bind_helper<concept_keyboard_device>(delegate_late_bind&) in libconcept.a[2](concept.o)
  "typeinfo for igs_xa_mcu_subcpu_device", referenced from:
      util::detail::delegate_generic_class* util::detail::delegate_late_bind_helper<delegate_late_bind>::late_bind_helper<igs_xa_mcu_subcpu_device>(delegate_late_bind&) in libigs.a[17](igs_m027xa.o)
      util::detail::delegate_generic_class* util::detail::delegate_late_bind_helper<delegate_late_bind>::late_bind_helper<igs_xa_mcu_subcpu_device>(delegate_late_bind&) in libigs.a[17](igs_m027xa.o)
  "vtable for device_finder<igs027a_cpu_device, true>", referenced from:
      std::__1::unique_ptr<device_t, std::__1::default_delete<device_t>> emu::detail::device_type_impl_base::create_driver<(anonymous namespace)::igs_fear_state>(emu::detail::device_type_impl_base const&, machine_config const&, char const*, device_t*, unsigned int) in libigs.a[14](igs_fear.o)
      std::__1::unique_ptr<device_t, std::__1::default_delete<device_t>> emu::detail::device_type_impl_base::create_driver<(anonymous namespace)::igs_m027_state>(emu::detail::device_type_impl_base const&, machine_config const&, char const*, device_t*, unsigned int) in libigs.a[15](igs_m027.o)
      std::__1::unique_ptr<device_t, std::__1::default_delete<device_t>> emu::detail::device_type_impl_base::create_driver<(anonymous namespace)::igs_m027xa_state>(emu::detail::device_type_impl_base const&, machine_config const&, char const*, device_t*, unsigned int) in libigs.a[17](igs_m027xa.o)
   NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "vtable for device_finder<igs_xa_mcu_subcpu_device, true>", referenced from:
      std::__1::unique_ptr<device_t, std::__1::default_delete<device_t>> emu::detail::device_type_impl_base::create_driver<(anonymous namespace)::igs_m027xa_state>(emu::detail::device_type_impl_base const&, machine_config const&, char const*, device_t*, unsigned int) in libigs.a[17](igs_m027xa.o)
   NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "vtable for device_finder<igs_xa_mcu_ics_sound_device, true>", referenced from:
      std::__1::unique_ptr<device_t, std::__1::default_delete<device_t>> emu::detail::device_type_impl_base::create_driver<(anonymous namespace)::igs_fear_state>(emu::detail::device_type_impl_base const&, machine_config const&, char const*, device_t*, unsigned int) in libigs.a[14](igs_fear.o)
   NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "_driver_tschess", referenced from:
      driver_list::s_drivers_sorted in drivlist.o
  "_driver_ultra45", referenced from:
      driver_list::s_drivers_sorted in drivlist.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [mame.make:270: ../../../../../mame] Error 1
make[1]: *** [Makefile:1135: mame] Error 2
make: *** [makefile:1349: macosx_x64_clang] Error 2`

I am already using the latest Xcode 15.4 Build version 15F31d

**Comment by cuavas** [comment]
You have existing generated project files, and the build system doesn’t automatically detect added files to regenerate the project files.  You can `touch makefile` or add `REGENIE=1` to force it to regenerate the project files.

**Reply by wyatt-wong** [reply]
> You have existing generated project files, and the build system doesn’t automatically detect added files to regenerate the project files. You can `touch makefile` or add `REGENIE=1` to force it to regenerate the project files.

Add REGENIE=1 to where ?

And which is to blame in this problem ? GitHub ? or the source code ? It would be nice if the makefile is automatically updated by itself. How would I know the makefile was OUTDATED ?

The basic approacah is always "git clone XXXX", "git pull" and "make"

**Comment by rb6502** [comment]
You should assume the makefile is outdated and use REGENIE=1 every time you pull new code.  We all do.

**Reply by wyatt-wong** [reply]
> You should assume the makefile is outdated and use REGENIE=1 every time you pull new code. We all do.

I didn't see it was mention in the build instruction.

What add REGENIE=1 to where ?

And when you said "we". what is it referrring ? Does "we" refers to the developers or contributators of MAME ? But I am neither of it. I simply want to build the latest macOS version of MAME since every new binary release did NOT include macOS.

**Reply by wyatt-wong** [reply]
Oh. I found this:

> 
drivers sources are added, removed, or renamed, the project files need
--
21 | to be regenerated.  You can do this by adding **REGENIE=1** to the
22 | make arguments, or updating the modification time of the makefile (for
 
27 | *chdman*, add a **TOOLS=1** to your make command, like
28 | **make REGENIE=1 TOOLS=1**
